.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,1174 @@
+        required: false
+        type: boolean
+        default: false # Always false by default, override on specific device types which this is relevant in the device repo
+      os-dev:
+        description: Enable OS development features
+        required: false
+        type: boolean
+        default: false # Only for use with manual runs/dispatches
+      deploy-esr:
+        description: Enable to deploy ESR
+        required: false
+        type: boolean
+        default: ${{ startsWith(github.event.push.ref, 'refs/tags/v20') }} # Set this flag if the tag looks like an esr tag - this is good for 100 years or until balenaOS v20 comes out
+      aws_iam_role:
+        description: IAM role to assume for s3 permissions
+        required: false
+        type: string
+        default: "arn:aws:iam::567579488761:role/balena-os-deploy-s3-access-Role-8r0SXqfzgolr" #TODO - Replace using environments
+      aws_region:
+        description: AWS region
+        required: false
+        type: string
+        default: "us-east-1"
+      test-suites:
+        description: "comma-separated list of test suites to run. No spaces in between."
+        required: false
+        type: string
+        default: "os,hup,cloud"
+      test-workers:
+        description: "Comma-separated list of worker types to use for testing. No spaces in between. Example - qemu or testbot."
+        required: false
+        type: string
+        default: "testbot" # Most devices use "testbot" , so qemu is the exception and should be set at the device repo level 
+      BALENACLOUD_APP_NAME:
+        description: "Testbot fleet for finding available Leviathan workers. Not used for QEMU workers. Can accept a list of apps separated by commas, no spaces in between."
+        type: string
+        required: false
+        default: "balena/testbot-rig,balena/testbot-rig-partners,balena/testbot-rig-x86,balena/testbot-rig-partners-x86"
+      BALENACLOUD_ORG: # Why is this an input and not a Var?
+        description: "It's weird, please remove it"
+        type: string
+        required: false
+        default: testbot
+
+# https://docs.github.com/en/actions/using-jobs/using-concurrency
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.number || github.ref }}-${{ inputs.machine }}
+  # cancel jobs in progress for updated PRs, but not merge or tag events
+  cancel-in-progress: ${{ github.event.action == 'synchronize' }}
+
+env:
+  WORKSPACE: ${{ github.workspace }}
+  MACHINE: ${{ inputs.machine }}
+  VERBOSE: verbose
+  WORKFLOW_NAME: ${{ github.workflow }} # Name of the calling workflow - for use when checking the result of test job on merged PR. Also, can we be clever here and also use it to differentiate between manual/auto runs
+
+permissions:
+  id-token: write   # This is required for requesting the JWT #https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#requesting-the-access-token
```

Comment:
Check which permissions we are losing by not specifying the default token permissions here.

Reply:
This is set to `write` for the sole purpose of getting the AWS token: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services#requesting-the-access-token

As far as I can tell its a hard requirement for this feature. This is the only thing we lose by not setting it to `write`

Reply:
Yes, but check this out: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token

> You can use the permissions key in your workflow file to modify permissions for the GITHUB_TOKEN for an entire workflow or for individual jobs. This allows you to configure the minimum required permissions for a workflow or job. When the permissions key is used, all unspecified permissions are set to no access, with the exception of the metadata scope, which always gets read access.

So by setting `id-token: write` and omitting all other permissions, we have removed all other permissions.

You can verify at the top of the `Set up job` step [here](https://github.com/balena-os/balena-yocto-scripts/actions/runs/9322094458/job/25663178018?pr=329).
```
Current runner version: '2.316.1'
Runner name: 'c3ccb3f8-079[2](https://github.com/balena-os/balena-yocto-scripts/actions/runs/9322094458/job/25663178018?pr=329#step:1:2)-4b8b-a4db-cf1cd3b04655'
Runner group name: 'self-hosted'
Machine name: 'f87[3](https://github.com/balena-os/balena-yocto-scripts/actions/runs/9322094458/job/25663178018?pr=329#step:1:3)2db'
GITHUB_TOKEN Permissions
  Metadata: read
Secret source: Actions
Prepare workflow directory
Prepare all required actions
Getting action download info
```

Instead we need [this](https://github.com/product-os/environment-production/blob/890f2e32975f7035511bb9da2fd9938e56f2155e/.github/workflows/flowzone.yml#L9-L25)

Reply:
Added `statuses: read` as I think thats the only extra permission we need for `GITHUB_TOKEN` - to read the status checks of the merge commits to see if tests passed. 

We can add other permissions as we discover we need them